### PR TITLE
ENH: Move factorizing for merge to EA interface

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jbrockmendel thoughts here? This would still need docs obviously

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -1703,6 +1704,15 @@ def _groupby_op(
         res_values = res_values.view(self._ndarray.dtype)
         return self._from_backing_data(res_values)
 
+    def _factorize_with_other(
+        self,
+        other: DatetimeLikeArrayMixin,  # type: ignore[override]
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:1709]: is other Self here?

File: pandas/core/arrays/base.py [hunk 2]
```diff
@@ -2266,6 +2285,13 @@ def _groupby_op(
         else:
             raise NotImplementedError
 
+    def _factorize_with_other(
+        self, other: ExtensionArray, sort: bool = False
+    ) -> tuple[npt.NDArray[np.intp], npt.NDArray[np.intp], int]:
+        lk, _ = self._values_for_factorize()
+        rk, _ = other._values_for_factorize()
+        return factorize_with_rizer(lk, rk, sort)
```
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:2293]: rizer is an awful name.  can we improve that while we're here?

File: pandas/core/reshape/merge.py [hunk 3]
```diff
@@ -2399,133 +2374,27 @@ def _factorize_keys(
     if (
         isinstance(lk.dtype, DatetimeTZDtype) and isinstance(rk.dtype, DatetimeTZDtype)
     ) or (lib.is_np_dtype(lk.dtype, "M") and lib.is_np_dtype(rk.dtype, "M")):
-        # Extract the ndarray (UTC-localized) values
         # Note: we dont need the dtypes to match, as these can still be compared
         lk, rk = cast("DatetimeArray", lk)._ensure_matching_resos(rk)
```
Comment (jbrockmendel) [inline pandas/core/reshape/merge.py:2378]: should the ensure_matching_resos be done in the EA method similar to how the Categorical one does encode_with_whatever?

File: pandas/core/arrays/base.py [hunk 4]
```diff
@@ -102,6 +104,23 @@
     from pandas import Index
 
 _extension_array_shared_docs: dict[str, str] = {}
+_factorizers = {
```
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:107]: looks like this isnt used in this file?

File: pandas/core/arrays/categorical.py [hunk 5]
```diff
@@ -2717,6 +2718,15 @@ def _groupby_op(
             res_values[result_mask == 1] = -1
         return self._from_backing_data(res_values)
 
+    def _factorize_with_other(
+        self, other: Categorical, sort: bool = False  # type: ignore[override]
+    ) -> tuple[npt.NDArray[np.intp], npt.NDArray[np.intp], int]:
+        other = self._encode_with_my_categories(other)
+
+        lk = ensure_int64(self.codes)
+        rk = ensure_int64(other.codes)
```
Comment (jbrockmendel) [inline pandas/core/arrays/categorical.py:2727]: is the ensure_int64 necessary?  im betting that is from a time before we had non-64 hashtables

File: pandas/core/arrays/categorical.py [hunk 6]
```diff
@@ -2717,6 +2718,15 @@ def _groupby_op(
             res_values[result_mask == 1] = -1
         return self._from_backing_data(res_values)
 
+    def _factorize_with_other(
+        self, other: Categorical, sort: bool = False  # type: ignore[override]
+    ) -> tuple[npt.NDArray[np.intp], npt.NDArray[np.intp], int]:
+        other = self._encode_with_my_categories(other)
```
Comment (jbrockmendel) [inline pandas/core/arrays/categorical.py:2724]: i think that factorize_with_other assumes matching dtypes. if correct, comment/docstring/assert to that effect?

File: pandas/core/reshape/merge.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/reshape/merge.py:2378]: i see, never mind

Comment (jbrockmendel) [reply pandas/core/reshape/merge.py:2378]: do we have tests with timedelta64 and mismatched resos?

File: pandas/core/arrays/base.py [hunk 7]
```diff
@@ -2266,6 +2285,13 @@ def _groupby_op(
         else:
             raise NotImplementedError
 
+    def _factorize_with_other(
```
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:2288]: maybe _factorize_with_other_for_merge?

File: pandas/core/reshape/merge_utils.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    llength = len(left)
+    labels = np.concatenate([left, right])
+
+    _, new_labels = algos.safe_sort(uniques, labels, use_na_sentinel=True)
+    new_left, new_right = new_labels[:llength], new_labels[llength:]
+
+    return new_left, new_right
+
+
+def factorize_with_rizer(lk, rk, sort: bool = False, lk_mask=None, rk_mask=None):
```
Comment (jbrockmendel) [inline pandas/core/reshape/merge_utils.py:45]: annotate?  it think lk/rk are ndarrays?

**Comment by jbrockmendel** [comment]
Generally looks good, mostly bikeshedding names

File: pandas/core/arrays/categorical.py (hunk 5 above)
Comment (phofl) [reply pandas/core/arrays/categorical.py:2727]: we run into trouble if they have different int dtypes, so casting both to int64 seems sensible

File: pandas/core/reshape/merge.py (hunk 3 above)
Comment (phofl) [reply pandas/core/reshape/merge.py:2378]: Don't know

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (phofl) [reply pandas/core/arrays/datetimelike.py:1709]: Oh yes, that's convenient

File: pandas/core/arrays/categorical.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/arrays/categorical.py:2727]: they should be matching dtypes though, right?

Comment (phofl) [reply pandas/core/arrays/categorical.py:2727]: simple reproducer:

```
ser = pd.Series([1, 2, 3, 1, 2], dtype="category")
cat = pd.Series(pd.Categorical.from_codes(ser.cat.codes.astype("int32"), ser.cat.categories))
assert ser.dtype == cat.dtype
```

dtype doesn't take the codes dtype into account

Comment (phofl) [reply pandas/core/arrays/categorical.py:2727]: We can also make this more efficient and do more extensive checking, but I think that we need something like this

Comment (jbrockmendel) [reply pandas/core/arrays/categorical.py:2727]: isnt that what encode_with_my_categories is for?  and even before that is called shouldn't we only get here with self.dtype == other.dtype?

Comment (phofl) [reply pandas/core/arrays/categorical.py:2727]: The dtypes are equal, the assertion passes.

```
_encode_with_my_categories
```
basically returns the same codes if the categories are equal

Comment (jbrockmendel) [reply pandas/core/arrays/categorical.py:2727]: i maintain the ensure_int64 shouldn't be necessary.  if you strongly disagree, can you add a TODO to follow up, pointing back to the relevant part of this thread

File: pandas/core/arrays/base.py (hunk 4 above)
Comment (phofl) [reply pandas/core/arrays/base.py:107]: Thx

File: pandas/core/arrays/base.py (hunk 7 above)
Comment (phofl) [reply pandas/core/arrays/base.py:2288]: Yeah that's probably better

File: pandas/core/reshape/merge_utils.py (hunk 8 above)
Comment (phofl) [reply pandas/core/reshape/merge_utils.py:45]: Added hints

File: pandas/core/arrays/categorical.py (hunk 5 above)
Comment (phofl) [reply pandas/core/arrays/categorical.py]: Sorry you are correct, removed the cast

File: pandas/core/arrays/arrow/array.py [hunk 9]
```diff
@@ -2553,6 +2555,47 @@ def _dt_tz_convert(self, tz):
...
+    ) -> tuple[npt.NDArray[np.intp], npt.NDArray[np.intp], int]:
+        if not isinstance(self.dtype, StringDtype) and (
+            pa.types.is_floating(self.dtype.pyarrow_dtype)
+            or pa.types.is_integer(self.dtype.pyarrow_dtype)
+            or pa.types.is_unsigned_integer(self.dtype.pyarrow_dtype)
+            or pa.types.is_duration(self.dtype.pyarrow_dtype)
+            or pa.types.is_date(self.dtype.pyarrow_dtype)
+            or pa.types.is_timestamp(self.dtype.pyarrow_dtype)
+        ):
+            lk = self.to_numpy(na_value=1, dtype=self.dtype.numpy_dtype)
```
Comment (jbrockmendel) [inline pandas/core/arrays/arrow/array.py:2569]: brief comment about why these particular cases are different?

File: pandas/core/arrays/arrow/array.py [hunk 10]
```diff
@@ -2553,6 +2555,47 @@ def _dt_tz_convert(self, tz):
...
+        left = self._pa_array
+        right = other._pa_array
+        dc = (
+            pa.chunked_array(left.chunks + right.chunks)
+            .combine_chunks()
+            .dictionary_encode()
+        )
+        length = len(dc.dictionary)
+
+        llab, rlab, count = (
```
Comment (jbrockmendel) [inline pandas/core/arrays/arrow/array.py:2588]: nitpick: can this be split into three lines.  it takes effort to see which lines correspond to which output

File: pandas/core/arrays/base.py [hunk 11]
```diff
@@ -2266,6 +2267,27 @@ def _groupby_op(
...
+        Parameters
+        ----------
+        other : ExtensionArray with the same dtype as self.
+        sort : Whether to sort the result.
+
+        Returns
+        -------
+        tuple of codes for left and right and the number of unique elements.
+        """
+        lk, _ = self._values_for_factorize()
```
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:2287]: assert/comment that dtypes are equal

File: pandas/core/arrays/datetimelike.py [hunk 12]
```diff
@@ -1703,6 +1704,13 @@ def _groupby_op(
         res_values = res_values.view(self._ndarray.dtype)
         return self._from_backing_data(res_values)
 
+    def _factorize_with_other_for_merge(
+        self, other: Self, sort: bool = False
+    ) -> tuple[npt.NDArray[np.intp], npt.NDArray[np.intp], int]:
+        lk = np.asarray(self._ndarray, dtype=np.int64)
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:1710]: `# caller is responsible for ensuring self.dtype == other.dtype`

**Comment by jbrockmendel** [comment]
a handful of nitpicks, otherwise im very happy to see this refactor

File: pandas/core/arrays/arrow/array.py (hunk 9 above)
Comment (phofl) [reply pandas/core/arrays/arrow/array.py:2569]: added

File: pandas/core/arrays/arrow/array.py (hunk 10 above)
Comment (phofl) [reply pandas/core/arrays/arrow/array.py:2588]: suer

File: pandas/core/arrays/base.py (hunk 11 above)
Comment (phofl) [reply pandas/core/arrays/base.py:2287]: added

File: pandas/core/arrays/datetimelike.py (hunk 12 above)
Comment (phofl) [reply pandas/core/arrays/datetimelike.py:1710]: added

**Comment by phofl** [comment]
Sorry this fell off my radar, updated

**Comment by phofl** [comment]
@MarcoGorelli is there a way to shut up the build about the missing example? There are a bunch of ea methods without examples...

**Comment by jbrockmendel** [comment]
> is there a way to shut up the build about the missing example? There are a bunch of ea methods without examples...

+1.  Many of the examples are not particularly useful (I know because I wrote them to get the CI to shut up)

File: ci/code_checks.sh [hunk 13]
```diff
@@ -1,3 +1,6 @@
+
+
+
```
Comment (jbrockmendel) [inline ci/code_checks.sh:3]: whitespace here intentional?

Comment (phofl) [reply ci/code_checks.sh]: no...

thx

**Comment by MarcoGorelli** [comment]
yeah I'd expect pasting into 

https://github.com/pandas-dev/pandas/blob/13e034aaf68894079c61ace35e2f9d541fd64a1e/scripts/validate_docstrings.py#L44-L58

to work

**Comment by lithomas1** [comment]
Looks like this has gone stale, so I'm bumping off the milestone

**Comment by mroeschke** [comment]
Looks close to merge but has gone stale. Going to close to clear the queue but feel free to reopen when you have time to circle back

**Pull request closed without merging**